File: core/src/query/optimizer.rs
//! Query plan optimizer.
//!
//! The [`Optimizer`] combines a [`RuleEngine`] with a set of built-in rules and
//! exposes a single [`Optimizer::optimize`] method that transforms a
//! [`LogicalPlan`] into an equivalent, lower-cost plan.
//!
//! ## Built-in rules (applied in order)
//!
//! 1. [`ConstantFolding`] — evaluate constant sub-expressions at plan time.
//! 2. [`PredicatePushdown`] — move filters towards leaf scans.
//! 3. [`ProjectionPushdown`] — prune unreferenced columns from scans.
//!
//! Join reordering based on cardinality estimates is performed separately in
//! [`reorder_joins`] and called as a post-pass after rule convergence.

use crate::query::logical_plan::{LogicalPlan, SortExpr};
use crate::query::rules::pushdown::{PredicatePushdown, ProjectionPushdown};
use crate::query::rules::{OptimizationRule, RuleEngine};
use crate::query::statistics::StatisticsRegistry;
use crate::sql::ast::{BinaryOperator, Expr, Value};

// ── ConstantFolding ───────────────────────────────────────────────────────────

/// Evaluates constant sub-expressions at plan time.
///
/// ### Supported rewrites
///
/// - `n + m` → `n+m` (integer arithmetic)
/// - `n * m` → `n*m` (integer arithmetic)
/// - `TRUE AND x` → `x`
/// - `FALSE AND x` → `FALSE`
/// - `TRUE OR x` → `TRUE`
/// - `FALSE OR x` → `x`
/// - `NOT TRUE` → `FALSE`
/// - `NOT FALSE` → `TRUE`
pub struct ConstantFolding;

impl OptimizationRule for ConstantFolding {
    fn name(&self) -> &str {
        "constant_folding"
    }

    fn apply(&self, plan: LogicalPlan) -> LogicalPlan {
        rewrite_plan_exprs(plan, fold_expr)
    }
}

/// Fold a single expression if it is a constant sub-expression.
fn fold_expr(expr: Expr) -> Expr {
    match expr {
        Expr::BinaryOp { left, op, right } => {
            let l = fold_expr(*left);
            let r = fold_expr(*right);
            fold_binary(l, op, r)
        }
        Expr::UnaryOp { op, expr } => {
            let inner = fold_expr(*expr);
            fold_unary(op, inner)
        }
        other => other,
    }
}

fn fold_binary(left: Expr, op: BinaryOperator, right: Expr) -> Expr {
    match (&left, &op, &right) {
        // Integer arithmetic
        (
            Expr::Literal(Value::Integer(a)),
            BinaryOperator::Plus,
            Expr::Literal(Value::Integer(b)),
        ) => Expr::Literal(Value::Integer(a + b)),
        (
            Expr::Literal(Value::Integer(a)),
            BinaryOperator::Minus,
            Expr::Literal(Value::Integer(b)),
        ) => Expr::Literal(Value::Integer(a - b)),
        (
            Expr::Literal(Value::Integer(a)),
            BinaryOperator::Multiply,
            Expr::Literal(Value::Integer(b)),
        ) => Expr::Literal(Value::Integer(a * b)),
        // Boolean short-circuit
        (Expr::Literal(Value::Boolean(true)), BinaryOperator::And, _) => right,
        (_, BinaryOperator::And, Expr::Literal(Value::Boolean(true))) => left,
        (Expr::Literal(Value::Boolean(false)), BinaryOperator::And, _) => {
            Expr::Literal(Value::Boolean(false))
        }
        (_, BinaryOperator::And, Expr::Literal(Value::Boolean(false))) => {
            Expr::Literal(Value::Boolean(false))
        }
        (Expr::Literal(Value::Boolean(true)), BinaryOperator::Or, _) => {
            Expr::Literal(Value::Boolean(true))
        }
        (_, BinaryOperator::Or, Expr::Literal(Value::Boolean(true))) => {
            Expr::Literal(Value::Boolean(true))
        }
        (Expr::Literal(Value::Boolean(false)), BinaryOperator::Or, _) => right,
        (_, BinaryOperator::Or, Expr::Literal(Value::Boolean(false))) => left,
        // No simplification possible
        _ => Expr::BinaryOp {
            left: Box::new(left),
            op,
            right: Box::new(right),
        },
    }
}

fn fold_unary(op: crate::sql::ast::UnaryOperator, expr: Expr) -> Expr {
    use crate::sql::ast::UnaryOperator;
    match (&op, &expr) {
        (UnaryOperator::Not, Expr::Literal(Value::Boolean(b))) => Expr::Literal(Value::Boolean(!b)),
        (UnaryOperator::Minus, Expr::Literal(Value::Integer(n))) => {
            Expr::Literal(Value::Integer(-n))
        }
        _ => Expr::UnaryOp {
            op,
            expr: Box::new(expr),
        },
    }
}

// ── Plan-level expression rewrite ─────────────────────────────────────────────

/// Walk the plan tree and apply `f` to every predicate / expression node.
fn rewrite_plan_exprs(plan: LogicalPlan, f: fn(Expr) -> Expr) -> LogicalPlan {
    match plan {
        LogicalPlan::Filter { input, predicate } => LogicalPlan::Filter {
            input: Box::new(rewrite_plan_exprs(*input, f)),
            predicate: f(predicate),
        },
        LogicalPlan::Scan {
            table,
            alias,
            columns,
            filter,
        } => LogicalPlan::Scan {
            table,
            alias,
            columns,
            filter: filter.map(f),
        },
        LogicalPlan::Project { input, items } => LogicalPlan::Project {
            input: Box::new(rewrite_plan_exprs(*input, f)),
            items,
        },
        LogicalPlan::Join {
            left,
            right,
            join_type,
            condition,
        } => LogicalPlan::Join {
            left: Box::new(rewrite_plan_exprs(*left, f)),
            right: Box::new(rewrite_plan_exprs(*right, f)),
            join_type,
            condition: condition.map(f),
        },
        LogicalPlan::Aggregate {
            input,
            group_by,
            aggregates,
            having,
        } => LogicalPlan::Aggregate {
            input: Box::new(rewrite_plan_exprs(*input, f)),
            group_by,
            aggregates,
            having: having.map(f),
        },
        LogicalPlan::Sort { input, order_by } => LogicalPlan::Sort {
            input: Box::new(rewrite_plan_exprs(*input, f)),
            order_by: order_by
                .into_iter()
                .map(|s| SortExpr {
                    expr: f(s.expr),
                    ascending: s.ascending,
                })
                .collect(),
        },
        LogicalPlan::Limit {
            input,
            limit,
            offset,
        } => LogicalPlan::Limit {
            input: Box::new(rewrite_plan_exprs(*input, f)),
            limit,
            offset,
        },
        LogicalPlan::Unnest {
            input,
            column,
            alias,
        } => LogicalPlan::Unnest {
            input: Box::new(rewrite_plan_exprs(*input, f)),
            column: f(column),
            alias,
        },
        LogicalPlan::ViewAs { input, items } => LogicalPlan::ViewAs {
            input: Box::new(rewrite_plan_exprs(*input, f)),
            items,
        },
        other => other,
    }
}

// ── Join reordering ───────────────────────────────────────────────────────────

/// Reorder a chain of inner joins by estimated cardinality (smallest first).
///
/// Only flat chains of `Inner` joins are reordered; other join types and
/// trees with non-trivial shapes are left unchanged to preserve semantics.
pub fn reorder_joins(plan: LogicalPlan, stats: &StatisticsRegistry) -> LogicalPlan {
    match plan {
        LogicalPlan::Join {
            left,
            right,
            join_type: crate::sql::ast::JoinType::Inner,
            condition,
        } => {
            let left_rows = scan_rows(&left, stats);
            let right_rows = scan_rows(&right, stats);
            if right_rows < left_rows {
                LogicalPlan::Join {
                    left: Box::new(reorder_joins(*right, stats)),
                    right: Box::new(reorder_joins(*left, stats)),
                    join_type: crate::sql::ast::JoinType::Inner,
                    condition,
                }
            } else {
                LogicalPlan::Join {
                    left: Box::new(reorder_joins(*left, stats)),
                    right: Box::new(reorder_joins(*right, stats)),
                    join_type: crate::sql::ast::JoinType::Inner,
                    condition,
                }
            }
        }
        LogicalPlan::Filter { input, predicate } => LogicalPlan::Filter {
            input: Box::new(reorder_joins(*input, stats)),
            predicate,
        },
        LogicalPlan::Sort { input, order_by } => LogicalPlan::Sort {
            input: Box::new(reorder_joins(*input, stats)),
            order_by,
        },
        LogicalPlan::Project { input, items } => LogicalPlan::Project {
            input: Box::new(reorder_joins(*input, stats)),
            items,
        },
        LogicalPlan::Aggregate {
            input,
            group_by,
            aggregates,
            having,
        } => LogicalPlan::Aggregate {
            input: Box::new(reorder_joins(*input, stats)),
            group_by,
            aggregates,
            having,
        },
        LogicalPlan::Limit {
            input,
            limit,
            offset,
        } => LogicalPlan::Limit {
            input: Box::new(reorder_joins(*input, stats)),
            limit,
            offset,
        },
        other => other,
    }
}

fn scan_rows(plan: &LogicalPlan, stats: &StatisticsRegistry) -> usize {
    match plan {
        LogicalPlan::Scan { table, .. } => stats.get(table).num_rows,
        _ => plan.estimated_rows(),
    }
}

// ── Optimizer ─────────────────────────────────────────────────────────────────

/// Entry point for logical plan optimization.
///
/// ```rust
/// use aeternumdb_core::query::optimizer::Optimizer;
/// use aeternumdb_core::query::statistics::StatisticsRegistry;
/// use aeternumdb_core::query::logical_plan::LogicalPlan;
///
/// let stats = StatisticsRegistry::new();
/// let optimizer = Optimizer::new(&stats);
/// let plan = LogicalPlan::Values { rows: vec![] };
/// let optimized = optimizer.optimize(plan);
/// ```
pub struct Optimizer<'a> {
    engine: RuleEngine,
    stats: &'a StatisticsRegistry,
}

impl<'a> Optimizer<'a> {
    /// Create a new optimizer with the default rule set.
    pub fn new(stats: &'a StatisticsRegistry) -> Self {
        let mut engine = RuleEngine::new();
        engine.add_rule(Box::new(ConstantFolding));
        engine.add_rule(Box::new(PredicatePushdown));
        engine.add_rule(Box::new(ProjectionPushdown));
        Self { engine, stats }
    }

    /// Optimize a logical plan.
    pub fn optimize(&self, plan: LogicalPlan) -> LogicalPlan {
        let rule_optimized = self.engine.optimize(plan);
        reorder_joins(rule_optimized, self.stats)
    }
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use crate::query::logical_plan::{LogicalPlan, ProjectionItem};
    use crate::sql::ast::{BinaryOperator, Expr, JoinType, UnaryOperator, Value};

    fn scan(t: &str) -> LogicalPlan {
        LogicalPlan::Scan {
            table: t.into(),
            alias: None,
            columns: None,
            filter: None,
        }
    }

    #[test]
    fn constant_fold_integer_add() {
        let rule = ConstantFolding;
        let expr = Expr::BinaryOp {
            left: Box::new(Expr::Literal(Value::Integer(2))),
            op: BinaryOperator::Plus,
            right: Box::new(Expr::Literal(Value::Integer(3))),
        };
        let plan = LogicalPlan::Filter {
            input: Box::new(scan("t")),
            predicate: expr,
        };
        let opt = rule.apply(plan);
        if let LogicalPlan::Filter { predicate, .. } = opt {
            assert_eq!(predicate, Expr::Literal(Value::Integer(5)));
        } else {
            panic!("expected Filter");
        }
    }

    #[test]
    fn constant_fold_true_and_x() {
        let rule = ConstantFolding;
        let x = Expr::Column {
            table: None,
            name: "active".into(),
        };
        let expr = Expr::BinaryOp {
            left: Box::new(Expr::Literal(Value::Boolean(true))),
            op: BinaryOperator::And,
            right: Box::new(x.clone()),
        };
        let plan = LogicalPlan::Filter {
            input: Box::new(scan("t")),
            predicate: expr,
        };
        let opt = rule.apply(plan);
        if let LogicalPlan::Filter { predicate, .. } = opt {
            assert_eq!(predicate, x);
        } else {
            panic!("expected Filter");
        }
    }

    #[test]
    fn constant_fold_not_true() {
        let rule = ConstantFolding;
        let expr = Expr::UnaryOp {
            op: UnaryOperator::Not,
            expr: Box::new(Expr::Literal(Value::Boolean(true))),
        };
        let plan = LogicalPlan::Filter {
            input: Box::new(scan("t")),
            predicate: expr,
        };
        let opt = rule.apply(plan);
        if let LogicalPlan::Filter { predicate, .. } = opt {
            assert_eq!(predicate, Expr::Literal(Value::Boolean(false)));
        } else {
            panic!("expected Filter");
        }
    }

    #[test]
    fn predicate_pushdown_applied_by_optimizer() {
        let mut reg = StatisticsRegistry::new();
        let mut ts = crate::query::statistics::TableStats::new("users");
        ts.num_rows = 100;
        reg.add(ts);

        let opt = Optimizer::new(&reg);
        let pred = Expr::BinaryOp {
            left: Box::new(Expr::Column {
                table: None,
                name: "age".into(),
            }),
            op: BinaryOperator::Gt,
            right: Box::new(Expr::Literal(Value::Integer(18))),
        };
        let plan = LogicalPlan::Filter {
            input: Box::new(scan("users")),
            predicate: pred,
        };
        let optimized = opt.optimize(plan);
        assert!(
            matches!(
                optimized,
                LogicalPlan::Scan {
                    filter: Some(_),
                    ..
                }
            ),
            "expected scan with pushed predicate"
        );
    }

    #[test]
    fn join_reorder_puts_smaller_table_first() {
        let mut reg = StatisticsRegistry::new();
        let mut big = crate::query::statistics::TableStats::new("orders");
        big.num_rows = 100_000;
        reg.add(big);
        let mut small = crate::query::statistics::TableStats::new("users");
        small.num_rows = 500;
        reg.add(small);

        let plan = LogicalPlan::Join {
            left: Box::new(scan("orders")),
            right: Box::new(scan("users")),
            join_type: JoinType::Inner,
            condition: None,
        };
        let result = reorder_joins(plan, &reg);
        if let LogicalPlan::Join { left, .. } = &result {
            assert!(
                matches!(left.as_ref(), LogicalPlan::Scan { table, .. } if table == "users"),
                "expected 'users' (smaller) on the left after reorder"
            );
        } else {
            panic!("expected Join");
        }
    }

    #[test]
    fn projection_pushdown_applied_by_optimizer() {
        let reg = StatisticsRegistry::new();
        let opt = Optimizer::new(&reg);
        let items = vec![ProjectionItem {
            expr: Expr::Column {
                table: None,
                name: "id".into(),
            },
            alias: None,
        }];
        let plan = LogicalPlan::Project {
            input: Box::new(scan("users")),
            items,
        };
        let optimized = opt.optimize(plan);
        if let LogicalPlan::Project { input, .. } = &optimized {
            assert!(
                matches!(
                    input.as_ref(),
                    LogicalPlan::Scan {
                        columns: Some(_),
                        ..
                    }
                ),
                "expected scan with column list"
            );
        } else {
            panic!("expected Project");
        }
    }
}

File: src/AeternumDB.Core/Query/Optimizer.cs
namespace AeternumDB.Core.Query;

using AeternumDB.Core.Query.Rules;
using AeternumDB.Core.Sql;

public sealed class ConstantFoldingRule : IOptimizationRule
{
    public string Name => "constant_folding";

    public LogicalPlan Apply(LogicalPlan plan) => RewritePlan(plan);

    private static LogicalPlan RewritePlan(LogicalPlan plan) =>
        plan switch
        {
            LogicalPlan.Filter f => new LogicalPlan.Filter(RewritePlan(f.Input), FoldExpr(f.Predicate)),
            LogicalPlan.Scan s => new LogicalPlan.Scan(s.Table, s.Alias, s.Columns, s.ScanFilter is null ? null : FoldExpr(s.ScanFilter)),
            LogicalPlan.Project p => new LogicalPlan.Project(
                RewritePlan(p.Input),
                [.. p.Items.Select(i => new ProjectionItem(FoldExpr(i.Expr), i.Alias))]),
            LogicalPlan.Join j => new LogicalPlan.Join(
                RewritePlan(j.Left),
                RewritePlan(j.Right),
                j.JoinType,
                j.Condition is null ? null : FoldExpr(j.Condition)),
            LogicalPlan.Aggregate a => new LogicalPlan.Aggregate(
                RewritePlan(a.Input),
                [.. a.GroupBy.Select(FoldExpr)],
                [.. a.Aggregates.Select(ae => new AggregateExpr(FoldExpr(ae.Func), ae.Alias))],
                a.Having is null ? null : FoldExpr(a.Having)),
            LogicalPlan.Sort s => new LogicalPlan.Sort(
                RewritePlan(s.Input),
                [.. s.OrderBy.Select(o => new SortExpr(FoldExpr(o.Expr), o.Ascending))]),
            LogicalPlan.Limit l => new LogicalPlan.Limit(RewritePlan(l.Input), l.LimitRows, l.OffsetRows),
            LogicalPlan.Unnest u => new LogicalPlan.Unnest(RewritePlan(u.Input), FoldExpr(u.Column), u.Alias),
            LogicalPlan.ViewAs v => new LogicalPlan.ViewAs(
                RewritePlan(v.Input),
                [.. v.Items.Select(i => new ViewAsProjection(FoldExpr(i.Expr), i.Alias))]),
            _ => plan
        };

    private static Expr FoldExpr(Expr expr)
    {
        if (expr is Expr.BinaryOp b)
        {
            var left = FoldExpr(b.Left);
            var right = FoldExpr(b.Right);
            return FoldBinary(left, b.Op, right);
        }

        if (expr is Expr.UnaryOp u)
        {
            var inner = FoldExpr(u.Inner);
            return FoldUnary(u.Op, inner);
        }

        return expr;
    }

    private static Expr FoldUnary(UnaryOperator op, Expr inner) =>
        (op, inner) switch
        {
            (UnaryOperator.Not, Expr.Literal { Value: SqlValue.Boolean b }) => new Expr.Literal(new SqlValue.Boolean(!b.Value)),
            (UnaryOperator.Minus, Expr.Literal { Value: SqlValue.Integer i }) => new Expr.Literal(new SqlValue.Integer(-i.Value)),
            _ => new Expr.UnaryOp(op, inner)
        };

    private static Expr FoldBinary(Expr left, BinaryOperator op, Expr right)
    {
        if (left is Expr.Literal { Value: SqlValue.Integer li } && right is Expr.Literal { Value: SqlValue.Integer ri })
            return FoldIntegerBinary(left, op, right, li.Value, ri.Value);

        if (left is Expr.Literal { Value: SqlValue.Boolean lb })
            return FoldBooleanLeft(lb.Value, op, right) ?? new Expr.BinaryOp(left, op, right);

        if (right is Expr.Literal { Value: SqlValue.Boolean rb })
            return FoldBooleanRight(left, op, rb.Value) ?? new Expr.BinaryOp(left, op, right);

        return new Expr.BinaryOp(left, op, right);
    }

    private static Expr FoldIntegerBinary(Expr left, BinaryOperator op, Expr right, long li, long ri) =>
        op switch
        {
            BinaryOperator.Plus => new Expr.Literal(new SqlValue.Integer(li + ri)),
            BinaryOperator.Minus => new Expr.Literal(new SqlValue.Integer(li - ri)),
            BinaryOperator.Multiply => new Expr.Literal(new SqlValue.Integer(li * ri)),
            _ => new Expr.BinaryOp(left, op, right)
        };

    private static Expr? FoldBooleanLeft(bool lb, BinaryOperator op, Expr right) =>
        (op, lb) switch
        {
            (BinaryOperator.And, true) => right,
            (BinaryOperator.And, false) => new Expr.Literal(new SqlValue.Boolean(false)),
            (BinaryOperator.Or, true) => new Expr.Literal(new SqlValue.Boolean(true)),
            (BinaryOperator.Or, false) => right,
            _ => null
        };

    private static Expr? FoldBooleanRight(Expr left, BinaryOperator op, bool rb) =>
        (op, rb) switch
        {
            (BinaryOperator.And, true) => left,
            (BinaryOperator.And, false) => new Expr.Literal(new SqlValue.Boolean(false)),
            (BinaryOperator.Or, true) => new Expr.Literal(new SqlValue.Boolean(true)),
            (BinaryOperator.Or, false) => left,
            _ => null
        };
}

public sealed class Optimizer(StatisticsRegistry stats)
{
    private readonly StatisticsRegistry _stats = stats;
    private readonly RuleEngine _engine = new RuleEngine()
        .AddRule(new ConstantFoldingRule())
        .AddRule(new PredicatePushdownRule())
        .AddRule(new ProjectionPushdownRule());

    public LogicalPlan Optimize(LogicalPlan plan)
    {
        var ruleOptimized = _engine.Optimize(plan);
        return ReorderJoins(ruleOptimized);
    }

    private LogicalPlan ReorderJoins(LogicalPlan plan) =>
        plan switch
        {
            LogicalPlan.Join j when j.JoinType == JoinType.Inner =>
                ReorderInnerJoin(j),
            LogicalPlan.Filter f => new LogicalPlan.Filter(ReorderJoins(f.Input), f.Predicate),
            LogicalPlan.Sort s => new LogicalPlan.Sort(ReorderJoins(s.Input), s.OrderBy),
            LogicalPlan.Project p => new LogicalPlan.Project(ReorderJoins(p.Input), p.Items),
            LogicalPlan.Aggregate a => new LogicalPlan.Aggregate(ReorderJoins(a.Input), a.GroupBy, a.Aggregates, a.Having),
            LogicalPlan.Limit l => new LogicalPlan.Limit(ReorderJoins(l.Input), l.LimitRows, l.OffsetRows),
            _ => plan
        };

    private LogicalPlan.Join ReorderInnerJoin(LogicalPlan.Join join)
    {
        // Keep explicit join predicates in-place to avoid invalidating
        // side-qualified column references when swapping inputs.
        if (join.Condition is not null)
            return new LogicalPlan.Join(ReorderJoins(join.Left), ReorderJoins(join.Right), JoinType.Inner, join.Condition);

        var leftRows = ScanRows(join.Left);
        var rightRows = ScanRows(join.Right);
        if (rightRows < leftRows)
            return new LogicalPlan.Join(ReorderJoins(join.Right), ReorderJoins(join.Left), JoinType.Inner, join.Condition);
        return new LogicalPlan.Join(ReorderJoins(join.Left), ReorderJoins(join.Right), JoinType.Inner, join.Condition);
    }

    private int ScanRows(LogicalPlan plan) =>
        plan switch
        {
            LogicalPlan.Scan s => _stats.Get(s.Table).NumRows,
            _ => plan.EstimatedRows()
        };
}
